pkg/activator/throttler.go

```diff
@@ -0,0 +1,176 @@
+func NewThrottler(queueDepth, maxConcurrency, initialCapacity int32, revisionInformer v1alpha1.RevisionInformer, logger *zap.SugaredLogger) *Throttler {
+	breakers := make(map[RevisionID]queue.Breaker)
+	return &Throttler{breakers: breakers, queueDepth: queueDepth, maxConcurrency: maxConcurrency, initialCapacity: initialCapacity, revisionInformer: revisionInformer, logger: logger}
+}
+
+// Throttler allows to update the max concurrency of registered Breakers
+type Throttler struct {
+	breakers         map[RevisionID]queue.Breaker
+	revisionInformer v1alpha1.RevisionInformer
+	queueDepth       int32
+	maxConcurrency   int32
+	initialCapacity  int32
+	logger           *zap.SugaredLogger
+	mux              sync.RWMutex
+}
+
+// Add potentially creates a new breaker for the revision
+// returns true if a new entry was added
+func (t *Throttler) AddIfAbsent(rev RevisionID) bool {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	if _, ok := t.breakers[rev]; !ok {
+		breaker := queue.NewBreaker(t.queueDepth, t.maxConcurrency, t.initialCapacity)
+		t.breakers[rev] = *breaker
+		return true
+	}
+	return false
+}
+
+// Exists returns true if a Breaker already exists for the given revision
+func (t *Throttler) Exists(revID RevisionID) bool {
+	t.mux.RLock()
+	defer t.mux.RUnlock()
+	_, ok := t.breakers[revID]
+	return ok
+}
+
+// Remove deletes the breaker from the bookkeeping
+func (t *Throttler) Remove(rev RevisionID) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	delete(t.breakers, rev)
+}
+
+// UpdateCapacity update the max concurrency of the Breaker corresponding to the revision
+func (t *Throttler) UpdateCapacity(rev RevisionID, size int32) error {
+	t.mux.RLock()
+	defer t.mux.RUnlock()
+	var containerConcurrency int32
+	revision, err := t.getRevision(rev.Namespace, rev.Name, t.revisionInformer)
+	if err != nil {
+		return fmt.Errorf("no revision was found for: %s", rev)
+	}
+	containerConcurrency = int32(revision.Spec.ContainerConcurrency)
+
+	if containerConcurrency == 0 {
+		containerConcurrency = t.maxConcurrency
+	}
+	if breaker, ok := t.breakers[rev]; ok {
+		err := breaker.UpdateConcurrency(size * containerConcurrency)
+		if err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("no breaker was found for the received revision: %s", rev)
+	}
+	return nil
+}
+
+// Try executes the `function` on the Breaker
+// Returns true if the function was executed, false otherwise
+// The function is not executed either if the Breaker doesn't exist or it is full
+func (t *Throttler) Try(rev RevisionID, function func()) error {
+	t.mux.RLock()
+	if breaker, ok := t.breakers[rev]; ok {
+		t.mux.RUnlock()
+		ok := breaker.Maybe(function)
+		if !ok {
+			return errors.New("activator throttler is in overflow")
+		}
+		return nil
+	}
+	t.mux.RUnlock()
+	return fmt.Errorf("breaker is not found for the revision %s", rev)
+}
+
+func (t *Throttler) getRevision(namespace, name string, informer v1alpha1.RevisionInformer) (*v1alpha12.Revision, error) {
```

Comment:
This doesn't seem to use anything in the Throttler. Does this need to belong to Throttler, or can we pull it out to just a function? I ask because I want to use this function to replace the existing logic on line 71 on revision.go for #1573.

Reply:
I think we can abstract it to just a function when we need it - would leave it in for the moment. According the #1573 - I had a similar plan to add a revision informer as part of getting rid of revision.go and deduper.go. Have you already started with the PR?